include/zephyr/devicetree/gpio.h

```diff
@@ -438,6 +438,198 @@ extern "C" {
+ * one element from @p "gpio-reserved-ranges"
+ */
+#define GEN_GPIO_RESERVED_RANGES_COND(idx, node_id)			   \
+	COND_CODE_1(DT_PROP_HAS_IDX(node_id, gpio_reserved_ranges, idx),   \
+		(GEN_PIN_MASK_COND(node_id, gpio_reserved_ranges,	   \
+			GET_ARG_N(idx, Z_SPARSE_LIST_EVEN_NUMBERS), idx)), \
+		(0))
+
+/**
+ * @brief Get a devicetree @p "gpio-reserved-ranges" property value
+ *
+ * This macro returns the value as a bitmask of the @p "gpio-reserved-ranges"
+ * property. This property defines the disabled (or 'reserved') GPIOs in the
+ * range @p 0...ngpios-1 and is specified as an array of value's pairs that
+ * define the start offset and size of the reserved ranges.
+ *
+ * For example, setting "gpio-reserved-ranges = <3 2>, <10 1>;"
+ * means that GPIO offsets 3, 4 and 10 cannot be used even if @p ngpios = <18>.
+ *
+ * The implementation constraint is inherited from common DT stuff: a maximum
+ * of 64 pairs can be used (with result limited to bitsize of gpio_port_pins_t
+ * type).
+ *
+ * NB: Due to the nature of C macros, some incorrect tuple definitions
+ *    (for example, overlapping or out of range) will produce undefined results.
+ *
+ *    Also be aware that if @p ngpios is less than 32 (bit size of DT int type),
+ *    then all unused MSBs outside the range defined by @p ngpios will be
+ *    marked as reserved too.
+ *
+ * Example devicetree fragment:
+ *
+ * @code{.dts}
+ *	a {
+ *		compatible = "some,gpio-controller";
+ *		ngpios = <32>;
+ *		gpio-reserved-ranges = <0  4>, <5  3>, <9  5>, <11 2>, <15 2>,
+ *					<18 2>, <21 1>, <23 1>, <25 4>, <30 2>;
+ *	};
+ *
+ *	b {
+ *		compatible = "some,gpio-controller";
+ *		ngpios = <18>;
+ *		gpio-reserved-ranges = <3 2>, <10 1>;
+ *	};
+ *
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *	struct some_config {
+ *		uint32_t ngpios;
+ *		uint32_t gpios_reserved;
+ *	};
+ *
+ *	static const struct some_config dev_cfg_a = {
+ *		.ngpios = DT_PROP_OR(DT_LABEL(a), ngpios, 0),
+ *		.gpios_reserved = DT_GPIO_RESERVED_RANGES(DT_LABEL(a)),
+ *	};
+ *
+ *	static const struct some_config dev_cfg_b = {
+ *		.ngpios = DT_PROP_OR(DT_LABEL(b), ngpios, 0),
+ *		.gpios_reserved = DT_GPIO_RESERVED_RANGES(DT_LABEL(b)),
+ *	};
+ *@endcode
+ *
+ * This expands to:
+ *
+ * @code{.c}
+ *	struct some_config {
+ *		uint32_t ngpios;
+ *		uint32_t gpios_reserved;
+ *	};
+ *
+ *	static const struct some_config dev_cfg_a = {
+ *		.ngpios = 32,
+ *		.gpios_reserved = 0xdeadbeef,
+ *		               // 0b1101 1110 1010 1101 1011 1110 1110 1111
+ *
+ *	static const struct some_config dev_cfg_b = {
+ *		.ngpios = 18,
+ *		.gpios_reserved = 0xfffc0418,
+ *		               // 0b1111 1111 1111 1100 0000 0100 0001 1000
+ *		               // unused MSBs were marked as reserved too
+ *	};
+ * @endcode
+ *
+ * @param node_id GPIO controller node identifier.
+ * @return the bitmask of reserved gpios
+ */
+#define DT_GPIO_RESERVED_RANGES(node_id)	((gpio_port_pins_t)	       \
```

Comment:
Hmm, no, I'm sorry to jump in late here but this is not acceptable to go into the devicetree API. I did not receive a review request until now for some reason.

The dependency layering is that gpio.h can depend on devicetree.h, but not the other way around. Please rework the PR so that anything which need to know anything about the Zephyr GPIO API is not in devicetree.h or any of its sub-headers like this one.

Reply:
> Hmm, no, I'm sorry to jump in late here but this is not acceptable to go into the devicetree API. I did not receive a review request until recently for some reason.

You are welcome, however somehow I have a feeling that my pebble caused an avalanche... :)

> 
> The dependency layering is that gpio.h can depend on devicetree.h, but not the other way around. Please rework the PR so that anything which need to know anything about the Zephyr GPIO API is not in devicetree.h or any of its sub-headers like this one.

Ah, I got it, but then where would it be better to put this things: into the **drivers/gpio.h** or return it back into **drivers/gpio/gpio_utils.h** ?

@talih0 IMO **drivers/gpio.h** is better suited for such type of thing as it's a missing part of the general GPIO API and not something ancillary as expected from the code placed in the **xxxx_utils.h** headers. However I don't know the subject's prehistory so suggestions are welcome.

@mbolivar-nordic should I also change the macro prefix from DT_GPIO_xx to GPIO_DT_xx, as is common in "drivers/gpio.h"? BTW it looks a bit confusing that we have two unrelated sets (DT_GPIO_ and GPIO_DT_) to describe parts of a whole. 
And as a consequence, is it acceptable to keep my tests in **tests/lib/devicetree/api**, or they have to relocated into the **tests/drivers/gpio/** ?